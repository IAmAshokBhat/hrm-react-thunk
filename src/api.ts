import {
  ILoginData,
  INewAttendance,
  INewLeave,
  IRegisterAPI
} from './constants';
import { fetchWithToken, getToken } from './utils';

const headers = {
  'X-JWT-Token': getToken()
};

export const getUserListAPI = () => fetchWithToken('user/list');

export const getUserDetailsAPI = (id: number) =>
  fetchWithToken(`api/users/${id}`);

export const getAttendanceDetailsAPI = (id: number, from: string, to: string) =>
  fetchWithToken(`api/attendance/${id}?from=${from}&to=${to}`);

export const getLeaveBalanceAPI = (id: number) =>
  fetchWithToken(`api/leaves/${id}`);

export const getEventDetailsAPI = (id: number, from: string, to: string) =>
  fetchWithToken(`api/events/${id}?from=${from}&to=${to}`);

export const getAllLeavesAPI = (id: number) =>
  fetchWithToken(`api/all-leaves/${id}`);

export const applyLeaveAPI = (newLeave: INewLeave) =>
  fetchWithToken(`api/apply-leave`, {
    method: 'POST',
    body: JSON.stringify(newLeave)
  });

export const addAttendanceAPI = (newAttendance: INewAttendance) =>
  fetchWithToken(`api/attendance`, {
    method: 'POST',
    body: JSON.stringify(newAttendance)
  });

export const loginAPI = (userInfo: ILoginData) =>
  fetchWithToken(`login`, {
    method: 'POST',
    body: JSON.stringify(userInfo)
  });

export const getDesignationListAPI = () =>
  fetchWithToken(`designation`, { headers: headers });

export const registerAPI = (newUser: IRegisterAPI) =>
  fetchWithToken(`user`, {
    method: 'POST',
    body: JSON.stringify(newUser)
  });
