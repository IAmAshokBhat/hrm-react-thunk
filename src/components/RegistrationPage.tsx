import Autocomplete from '@mui/material/Autocomplete';
import Button from '@mui/material/Button';
import Card from '@mui/material/Card';
import FormControl from '@mui/material/FormControl';
import Grid from '@mui/material/Grid';
import TextField from '@mui/material/TextField';
import Typography from '@mui/material/Typography';
import { useEffect, useState } from 'react';
import { Field, Form } from 'react-final-form';
import { useDispatch, useSelector } from 'react-redux';
import { NotificationType } from '../constants';
import { ILabelValue, IRegisterAPI, IRegistration } from '../contracts';
import {
  fetchDesignationListAction,
  fetchUsersAction,
  registerAction
} from '../redux/actions';
import {
  selectDesignationList,
  selectUserList,
  selectRegister
} from '../redux/selectors';
import { useSnackbar } from 'notistack';
import { useNavigate } from 'react-router-dom';

export const RegistrationPage = () => {
  const dispatch = useDispatch();

  const navigate = useNavigate();

  const { enqueueSnackbar } = useSnackbar();

  const [selectedDesignation, setDesignation] = useState<ILabelValue | null>(
    null
  );
  const [selectedManager, setManger] = useState<ILabelValue | null>(null);

  const designationList = useSelector(selectDesignationList);

  const mangerList = useSelector(selectUserList);

  const registrationResponse = useSelector(selectRegister);

  const onSubmit = (values: IRegistration) => {
    if (selectedManager && selectedDesignation) {
      const { name, password, email } = values;
      const newUser: IRegisterAPI = {
        name,
        password,
        email,
        manager: selectedManager.value,
        designation: selectedDesignation.value
      };
      dispatch(registerAction(newUser));
    }
  };

  useEffect(() => {
    dispatch(fetchDesignationListAction());
    dispatch(fetchUsersAction());
    if (registrationResponse.status) {
      if (registrationResponse.status === 200) {
        enqueueSnackbar('User added successfully!', {
          variant: NotificationType.SUCCESS
        });
        navigate('/login');
      } else {
        enqueueSnackbar(registrationResponse.message, {
          variant: NotificationType.ERROR
        });
      }
    }
  }, [dispatch, registrationResponse, enqueueSnackbar, navigate]);

  return (
    <Card style={{ maxWidth: 500, margin: '0 auto', padding: 20 }}>
      <Form
        onSubmit={onSubmit}
        render={({ handleSubmit, values, dirtyFields }) => {
          const isPasswordNotMatching =
            values.password !== values.confirmPassword;
          const isFormValid =
            !isPasswordNotMatching &&
            selectedManager !== null &&
            selectedDesignation !== null;
          return (
            <form onSubmit={handleSubmit}>
              <Grid container rowSpacing={2}>
                <Grid item xs={12}>
                  <Typography textAlign="center" variant="h4">
                    HR Management System
                  </Typography>
                </Grid>
                <Grid item xs={12}>
                  <FormControl sx={{ width: '100%' }}>
                    <Field name="name" width="100%">
                      {(props) => (
                        <TextField
                          name={props.input.name}
                          value={props.input.value}
                          onChange={props.input.onChange}
                          label="User Name"
                          variant="outlined"
                          placeholder="User Name"
                        />
                      )}
                    </Field>
                  </FormControl>
                </Grid>
                <Grid item xs={12}>
                  <FormControl sx={{ width: '100%' }}>
                    <Field name="email" width="100%">
                      {(props) => (
                        <TextField
                          name={props.input.name}
                          value={props.input.value}
                          onChange={props.input.onChange}
                          label="Email"
                          placeholder="Email"
                          variant="outlined"
                        />
                      )}
                    </Field>
                  </FormControl>
                </Grid>
                <Grid item xs={12}>
                  <FormControl sx={{ width: '100%' }}>
                    {designationList && (
                      <Autocomplete
                        id="leave-type"
                        value={selectedDesignation}
                        isOptionEqualToValue={(option, value) =>
                          option.value === value.value
                        }
                        options={designationList}
                        onChange={(_event, newValue: ILabelValue | null) =>
                          setDesignation(newValue)
                        }
                        renderInput={(params) => (
                          <TextField
                            {...params}
                            label="Designation"
                            variant="outlined"
                            placeholder="Designation"
                          />
                        )}
                      />
                    )}
                  </FormControl>
                </Grid>
                <Grid item xs={12}>
                  <FormControl sx={{ width: '100%' }}>
                    {mangerList && (
                      <Autocomplete
                        id="manager"
                        value={selectedManager}
                        options={mangerList}
                        isOptionEqualToValue={(option, value) =>
                          option.value === value.value
                        }
                        onChange={(_event, newValue: ILabelValue | null) =>
                          setManger(newValue)
                        }
                        renderInput={(params) => (
                          <TextField
                            {...params}
                            label="Manager"
                            variant="outlined"
                            placeholder="Manager"
                          />
                        )}
                      />
                    )}
                  </FormControl>
                </Grid>
                <Grid item xs={12}>
                  <FormControl sx={{ width: '100%' }}>
                    <Field name="password">
                      {(props) => (
                        <TextField
                          name={props.input.name}
                          value={props.input.value}
                          onChange={props.input.onChange}
                          fullWidth
                          label="Password"
                          variant="outlined"
                          type="password"
                          placeholder="Password"
                          inputProps={{ 'data-testid': 'password' }}
                        />
                      )}
                    </Field>
                  </FormControl>
                </Grid>
                <Grid item xs={12}>
                  <FormControl sx={{ width: '100%' }}>
                    <Field name="confirmPassword">
                      {(props) => (
                        <TextField
                          name={props.input.name}
                          value={props.input.value}
                          onChange={props.input.onChange}
                          fullWidth
                          label="Confirm Password"
                          variant="outlined"
                          error={
                            dirtyFields.confirmPassword && isPasswordNotMatching
                          }
                          type="password"
                          placeholder="Confirm Password"
                        />
                      )}
                    </Field>
                  </FormControl>
                </Grid>
                <Grid item xs={12} textAlign="right">
                  <Button
                    variant="contained"
                    type="submit"
                    fullWidth
                    disabled={!isFormValid}
                  >
                    Register
                  </Button>
                </Grid>
              </Grid>
            </form>
          );
        }}
      />
    </Card>
  );
};
