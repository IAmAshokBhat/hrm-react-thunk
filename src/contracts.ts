import { AnyAction, Dispatch } from 'redux';
import { TAny, TAnyObject } from './constants';

export type TFPromise = (params: TAny) => Promise<TAnyObject>;

export type TFPromiseAPI = (params?: TAny | TAnyObject) => Promise<TAnyObject>;

export type TFAction = (dispatch: Dispatch<AnyAction>) => Promise<void>;

/*
()
(1,2,3)
({key:value})
**
*/
