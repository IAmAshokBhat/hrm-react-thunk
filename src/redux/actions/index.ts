import { fetchUserDetailsAction } from './userInfoAction';
import { fetchUsersListAction } from './usersActions';
import {
  fetchAttendanceDetailsAction,
  fetchAllLeavesAction
} from './attedanceDetailsAction';
import { fetchLeaveBalanceAction } from './leaveBalanceAction';
import { fetchEventDetailsAction } from './eventDetailsAction';
import { applyLeaveAction } from './applyLeaveAction';
import { addAttendanceAction } from './addAttendanceAction';
import { loginAction } from './loginAction';
import { registerAction } from './registerAction';
import { fetchDesignationListAction } from './designationListAction';

export {
  fetchUsersListAction as fetchUsersAction,
  fetchUserDetailsAction,
  fetchAttendanceDetailsAction,
  fetchLeaveBalanceAction,
  fetchEventDetailsAction,
  applyLeaveAction,
  addAttendanceAction,
  loginAction,
  registerAction,
  fetchDesignationListAction,
  fetchAllLeavesAction
};
