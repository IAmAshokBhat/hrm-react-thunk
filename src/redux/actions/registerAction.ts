import { Dispatch } from 'redux';
import { registerAPI } from '../../api';
import { ACTIONS, IGenericResponse, IRegisterAPI } from '../../constants';

const setRegisterFailureAction = () => ({
  type: ACTIONS.REGISTER_FAILURE
});

const setRegisterLoadingAction = () => ({
  type: ACTIONS.REGISTER_LOADING
});

const registerSuccessAction = (payload: IGenericResponse) => ({
  type: ACTIONS.REGISTER_SUCCESS,
  payload
});

export const registerAction =
  (newUser: IRegisterAPI) => (dispatch: Dispatch) => {
    dispatch(setRegisterLoadingAction());
    return registerAPI(newUser)
      .then(
        (response) => response.json(),
        (error) => console.log('An error occurred.', error)
      )
      .then((json) => {
        const { ok } = json;
        if (!ok) {
          dispatch(setRegisterFailureAction());
        }
        dispatch(registerSuccessAction(json));
      });
  };
