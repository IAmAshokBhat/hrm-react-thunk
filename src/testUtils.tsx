import { render, RenderOptions, waitFor } from '@testing-library/react';
import { SnackbarProvider } from 'notistack';
import { ReactElement } from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { applyMiddleware, createStore, Store } from 'redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { IDefaultRoot, TAny, TAnyObject } from './constants';
import { TFAction } from './contracts';
import { reducers } from './redux/reducers';
import {
  defaultArrayState,
  defaultNullState,
  defaultState
} from './redux/reducers/__tests__/fixtures';

export const middlewares = [thunk];

export const createStoreWithMiddlewares = (
  initialState: TAnyObject = {}
): Store => {
  return createStore(reducers, initialState, applyMiddleware(...middlewares));
};
interface IRenderWithProviderOptions {
  initialStoreState?: TAnyObject;
  renderFunction?: TAny;
  options?: RenderOptions;
  route?: string;
  path?: string;
}

export const renderWithProvider = (
  ui: ReactElement,
  renderOptions?: IRenderWithProviderOptions
) => {
  const {
    initialStoreState = {},
    renderFunction = render,
    options = {},
    route = '',
    path = ''
  } = renderOptions || {};

  const appStore = createStoreWithMiddlewares(initialStoreState);

  let wrappedUi = (
    <SnackbarProvider
      anchorOrigin={{
        vertical: 'top',
        horizontal: 'right'
      }}
    >
      <Provider store={appStore}>{ui}</Provider>
    </SnackbarProvider>
  );

  if (route) {
    window.history.pushState({}, '', route);

    wrappedUi = (
      <Provider store={appStore}>
        <BrowserRouter>
          <Routes>
            <Route path={path} element={ui} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );
  }

  return renderFunction(wrappedUi, options);
};

type TFApiCall = (...params: TAny) => Promise<Response>;

export interface IResponseMock {
  headers?: object;
  ok?: boolean;
  status?: number;
  statusText?: string;
  type?: ResponseType;
  url?: string;
  body?: TAny | null;
  bodyUsed?: boolean;
  blob?: () => Promise<TAny>;
}

const dummyPromise = () => new Promise((resolve) => resolve('dummy'));

export class ResponseMock {
  headers: object;

  ok: boolean;

  status: number;

  statusText: string;

  type: ResponseType;

  url: string;

  body: ReadableStreamDefaultReadResult<Uint8Array> | null;

  bodyUsed: boolean;

  blob: () => Promise<TAny>;

  constructor(props: IResponseMock = {}) {
    this.headers = props.headers || {};
    this.status = props.status || 200;
    this.ok = props.ok || this.status === 200;
    this.statusText = props.statusText || 'OK';
    this.type = props.type || 'basic';
    this.url = props.url || 'http://localhost';
    this.body = props.body;
    this.bodyUsed = props.bodyUsed || this.body !== null;
    this.blob = props.blob || dummyPromise;
  }

  json: () => Promise<TAny> = () => {
    const body = this.body || '""';

    try {
      return Promise.resolve(JSON.parse(body as string));
    } catch {
      return Promise.resolve(body);
    }
  };

  text: () => Promise<string> = () => Promise.resolve(this.statusText);
}

export const mockAPICall = (func: TFApiCall, response: IResponseMock) =>
  (func as jest.Mock).mockResolvedValue(new ResponseMock(response));

export const generateActionTests = async (
  actionName: string,
  action: TFAction,
  api: TAny,
  actionPayload: TAny = []
) => {
  const response = new ResponseMock({ ok: true, body: { data: [] } });

  const mockStore = configureStore(middlewares);

  (api as jest.Mock).mockReturnValue(Promise.resolve(response));
  const store = mockStore({});

  //TODO: Fix TS error
  // @ts-ignore
  await store.dispatch(action);

  await waitFor(() => {
    expect(store.getActions()).toEqual([
      {
        type: `${actionName}_LOADING`
      },
      {
        type: `${actionName}_SUCCESS`,
        //TODO: Accept action params from function params
        payload: actionPayload
      }
    ]);
  });
};

export const generateReducerTests = (
  reducerName: string,
  reducer: TAny,
  selector: TAny,
  actionName: string,
  successValue?: TAny,
  defaultValue?: TAny,
  expectedValue?: TAny,
  reducerLabel?: string
) => {
  describe(`for ${reducerName}`, () => {
    const reducerLbl = reducerLabel
      ? reducerLabel
      : reducerName.replace('Reducer', '');
    it(`${reducerName} reducer loading`, () => {
      reducer(undefined, {
        type: `${actionName}_LOADING`,
        payload: null
      });
      // TODO: Check why its failing
      // expect(selector(defaultState).loading).toEqual(true);
      expect(selector(defaultState).value).toEqual(
        defaultValue === null ? defaultValue : []
      );
      expect(selector(defaultState).error).toEqual(null);
    });

    it(`${reducerName} reducer success`, () => {
      const successState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: expectedValue ? expectedValue : successValue
        }
      };
      reducer(defaultArrayState, {
        type: `${actionName}_SUCCESS`,
        payload: successValue
      });

      expect(selector(successState).error).toEqual(null);
      expect(selector(successState).loading).toEqual(false);
      expect(selector(successState).value).toEqual(
        expectedValue ? expectedValue : successValue
      );
    });

    it(`${reducerName} reducer error`, () => {
      const error = { message: '', status: 1, ok: false };
      const failedState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: [],
          error
        }
      };

      reducer(undefined, {
        type: `${actionName}_FAILURE`,
        payload: error
      });

      expect(selector(failedState).loading).toEqual(false);
      expect(selector(failedState).value).toEqual([]);
      expect(selector(failedState).error).toEqual(error);
    });
  });
};
