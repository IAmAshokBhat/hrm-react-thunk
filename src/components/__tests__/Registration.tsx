import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SnackbarProvider } from 'notistack';
import { getDesignationListAPI, getUserListAPI, registerAPI } from '../../api';
import { TAnyObject } from '../../contracts';
import * as Actions from '../../redux/actions/registerAction';
import { mockAPICall, renderWithProvider } from '../../testUtils';
import { RegistrationPage } from '../RegistrationPage';
import {
  defaultState,
  designationListAPIResponse,
  userListState
} from './fixtures';

jest.mock('../../api', () => ({
  registerAPI: jest.fn(),
  getDesignationListAPI: jest.fn(),
  getUserListAPI: jest.fn()
}));

const mockEnqueue = jest.fn();

jest.mock('notistack', () => ({
  ...jest.requireActual('notistack'),
  useSnackbar: () => {
    return {
      enqueueSnackbar: mockEnqueue
    };
  }
}));

describe('Registration Page', () => {
  beforeEach(() => {
    mockAPICall(registerAPI, {
      ok: true,
      body: { data: [], message: 'success', status: 200 }
    });
    mockAPICall(getDesignationListAPI, {
      ok: true,
      body: designationListAPIResponse
    });
    mockAPICall(getUserListAPI, {
      ok: true,
      body: userListState
    });
  });

  const generateNode = (initialStoreState: TAnyObject = defaultState) =>
    renderWithProvider(
      <SnackbarProvider>
        <RegistrationPage />
      </SnackbarProvider>,
      {
        route: `/register`,
        path: '/register',
        initialStoreState
      }
    );

  it('should render and match snapshot', () => {
    const { container } = generateNode();

    expect(container).toMatchSnapshot();
  });

  it('should be able to register', async () => {
    const registerActionSpy = jest.spyOn(Actions, 'registerAction');
    generateNode({
      ...defaultState,
      designationList: {
        error: null,
        loading: false,
        value: designationListAPIResponse.data
      },
      userList: { error: null, loading: false, value: userListState }
    });
    const registerBtn = screen.getByRole('button', { name: 'Register' });

    expect(registerBtn).toBeDisabled();
    userEvent.type(screen.getByPlaceholderText(/User Name/i), 'test name');
    userEvent.type(screen.getByPlaceholderText(/Email/i), 'test email');
    userEvent.click(
      screen.getByRole('textbox', {
        name: /designation/i
      })
    );
    userEvent.click(screen.getByRole('option', { name: 'Vice President' }));
    userEvent.click(
      screen.getByRole('textbox', {
        name: /manager/i
      })
    );
    userEvent.click(screen.getByRole('option', { name: 'x' }));
    const passwordEl = screen.getByTestId('password');
    await waitFor(() => userEvent.type(passwordEl, 'p'));
    expect(passwordEl).toHaveValue('p');
    const confirmPasswordEl = screen.getByLabelText(/confirm password/i);
    userEvent.type(confirmPasswordEl, 'd');
    expect(confirmPasswordEl).toHaveAttribute('aria-invalid', 'true');
    await waitFor(() => userEvent.clear(confirmPasswordEl));
    await waitFor(() => userEvent.type(confirmPasswordEl, 'p'));
    expect(confirmPasswordEl).toHaveValue('p');
    expect(confirmPasswordEl).toHaveAttribute('aria-invalid', 'false');

    await waitFor(() => userEvent.click(registerBtn));

    await waitFor(() =>
      expect(registerActionSpy).toHaveBeenCalledWith({
        designation: 14,
        email: 'test email',
        manager: 1,
        name: 'test name',
        password: 'p'
      })
    );

    await waitFor(() =>
      expect(registerAPI).toHaveBeenCalledWith({
        designation: 14,
        email: 'test email',
        manager: 1,
        name: 'test name',
        password: 'p'
      })
    );

    expect(mockEnqueue).toHaveBeenCalled();
  });
});
