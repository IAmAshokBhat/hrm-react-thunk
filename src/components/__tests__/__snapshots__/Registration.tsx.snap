// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration Page should render and match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    style="max-width: 500px; margin: 0px auto; padding: 20px;"
  >
    <form>
      <div
        class="MuiGrid-root MuiGrid-container css-1jzi3sl-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-149xwtl-MuiTypography-root"
          >
            HR Management System
          </h4>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="mui-1"
                id="mui-1-label"
              >
                User Name
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-1"
                  name="name"
                  placeholder="User Name"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      User Name
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="mui-2"
                id="mui-2-label"
              >
                Email
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-2"
                  name="email"
                  placeholder="Email"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      Email
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
          >
            <div
              aria-expanded="false"
              class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-16awh2u-MuiAutocomplete-root"
              role="combobox"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="leave-type"
                  id="leave-type-label"
                >
                  Designation
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-autocomplete="list"
                    aria-invalid="false"
                    autocapitalize="none"
                    autocomplete="off"
                    class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                    id="leave-type"
                    placeholder="Designation"
                    spellcheck="false"
                    type="text"
                    value=""
                  />
                  <div
                    class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                  >
                    <button
                      aria-label="Open"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                      tabindex="-1"
                      title="Open"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="ArrowDropDownIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-173wfxe"
                    >
                      <span>
                        Designation
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
          >
            <div
              aria-expanded="false"
              class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-16awh2u-MuiAutocomplete-root"
              role="combobox"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="manager"
                  id="manager-label"
                >
                  Manager
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-autocomplete="list"
                    aria-invalid="false"
                    autocapitalize="none"
                    autocomplete="off"
                    class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                    id="manager"
                    placeholder="Manager"
                    spellcheck="false"
                    type="text"
                    value=""
                  />
                  <div
                    class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                  >
                    <button
                      aria-label="Open"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                      tabindex="-1"
                      title="Open"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="ArrowDropDownIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-173wfxe"
                    >
                      <span>
                        Manager
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="mui-3"
                id="mui-3-label"
              >
                Password
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="password"
                  id="mui-3"
                  name="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      Password
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="mui-4"
                id="mui-4-label"
              >
                Confirm Password
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-4"
                  name="confirmPassword"
                  placeholder="Confirm Password"
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      Confirm Password
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13ezsq2-MuiGrid-root"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root Mui-disabled  css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            Register
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
